File: hakimi-erp/backend/app/schemas/logistics.py
from typing import Optional, List
from pydantic import BaseModel, model_validator
from datetime import datetime, date
from decimal import Decimal

# --- Delivery Item ---
class DeliveryItemBase(BaseModel):
    item_no: int
    so_item_id: Optional[str] = None
    material_id: str
    order_quantity: Optional[Decimal] = None
    delivery_quantity: Optional[Decimal] = None
    picked_quantity: Optional[Decimal] = None
    sales_unit: Optional[str] = None
    plant: Optional[str] = None
    storage_location: Optional[str] = None
    item_description: Optional[str] = None
    item_status: Optional[str] = "OPEN"

class DeliveryItemCreate(DeliveryItemBase):
    delivery_item_id: str

class DeliveryItem(DeliveryItemBase):
    delivery_item_id: str
    delivery_id: str
    material_name: Optional[str] = None
    base_unit: Optional[str] = None

    class Config:
        from_attributes = True

# --- Delivery ---
class DeliveryBase(BaseModel):
    delivery_id: str
    sales_order_id: Optional[str] = None
    delivery_type: str
    delivery_status: str
    ship_to_party: str
    planned_delivery_date: Optional[date] = None
    planned_gi_date: Optional[date] = None
    actual_gi_date: Optional[datetime] = None
    picking_date: Optional[date] = None
    shipping_point: Optional[str] = None
    carrier: Optional[str] = None
    driver_name: Optional[str] = None
    route: Optional[str] = None
    tracking_no: Optional[str] = None

class DeliveryCreate(DeliveryBase):
    items: List[DeliveryItemCreate]

class Delivery(DeliveryBase):
    ship_to_party_name: Optional[str] = None
    ship_to_address: Optional[str] = None
    items: List[DeliveryItem] = []
    total_quantity: Optional[Decimal] = None
    delivered_quantity: Optional[Decimal] = None

    class Config:
        from_attributes = True

# --- Delivery List Item (lightweight) ---
class DeliveryListItem(BaseModel):
    delivery_id: str
    sales_order_id: Optional[str] = None
    delivery_type: str
    delivery_status: str
    ship_to_party: str
    ship_to_party_name: Optional[str] = None
    planned_delivery_date: Optional[date] = None
    planned_gi_date: Optional[date] = None
    actual_gi_date: Optional[datetime] = None
    picking_date: Optional[date] = None
    shipping_point: Optional[str] = None
    total_quantity: Optional[Decimal] = None
    delivered_quantity: Optional[Decimal] = None

    class Config:
        from_attributes = True

# --- Goods Issue ---
class GoodsIssueBase(BaseModel):
    goods_issue_id: str
    delivery_item_id: str
    actual_quantity: Decimal
    posting_date: date
    warehouse: Optional[str] = None
    batch_no: Optional[int] = 1

class GoodsIssueCreate(GoodsIssueBase):
    pass

class GoodsIssue(GoodsIssueBase):
    goods_issue_time: datetime
    material_id: Optional[str] = None
    material_name: Optional[str] = None

    class Config:
        from_attributes = True


# --- Partial Picking / PGI request schemas ---
class PartialItemQuantity(BaseModel):
    delivery_item_id: str
    quantity: Decimal

class PartialPickingRequest(BaseModel):
    items: List[PartialItemQuantity]

class PartialPgiRequest(BaseModel):
    items: List[PartialItemQuantity]


# --- Pick Record (batch picking) ---
class PickRecordBase(BaseModel):
    pick_id: str
    delivery_item_id: str
    batch_no: int
    pick_quantity: Decimal
    storage_location: Optional[str] = None
    pick_date: datetime
    picked_by: Optional[str] = None

class PickRecord(PickRecordBase):
    material_id: Optional[str] = None
    material_name: Optional[str] = None

    class Config:
        from_attributes = True

class PickBatchRequest(BaseModel):
    items: List[PartialItemQuantity]
    storage_location: Optional[str] = None
    picked_by: Optional[str] = None

# --- SAP-style remaining quantity ---
class SoItemRemaining(BaseModel):
    so_item_id: str
    material_id: str
    material_name: Optional[str] = None
    order_quantity: Decimal
    already_delivered: Decimal
    remaining_quantity: Decimal
    sales_unit: Optional[str] = None

# --- Status mapping helpers ---
DELIVERY_STATUS_LABELS = {
    "OPEN": "Creating",
    "PICKING": "Picking",
    "SHIPPED": "Picked",
    "IN_TRANSIT": "In Transit",
    "PGI_DONE": "Completed",
    "CANCELLED": "Cancelled",
}

def get_status_label(status: str) -> str:
    return DELIVERY_STATUS_LABELS.get(status, status)

def get_status_index(status: str) -> int:
    mapping = {"OPEN": 0, "PICKING": 1, "SHIPPED": 2, "IN_TRANSIT": 3, "PGI_DONE": 4, "CANCELLED": -1}
    return mapping.get(status, 0)

File: hakimi-erp/frontend/src/api/modules/logistics.ts
import { get, post, put } from '../request'

// --- Delivery Item ---
export interface DeliveryItem {
  deliveryItemId: string
  deliveryId: string
  itemNo: number
  soItemId?: string
  materialId: string
  materialName?: string
  baseUnit?: string
  orderQuantity?: number
  deliveryQuantity: number
  pickedQuantity?: number
  salesUnit?: string
  plant?: string
  storageLocation?: string
  itemDescription?: string
  itemStatus?: string
}

// --- Delivery (Detail) ---
export interface Delivery {
  deliveryId: string
  salesOrderId?: string
  deliveryType: string
  deliveryStatus: string
  shipToParty: string
  shipToPartyName?: string
  shipToAddress?: string
  plannedDeliveryDate?: string
  plannedGiDate?: string
  actualGiDate?: string
  pickingDate?: string
  shippingPoint?: string
  carrier?: string
  driverName?: string
  route?: string
  trackingNo?: string
  totalQuantity?: number
  deliveredQuantity?: number
  items: DeliveryItem[]
}

// --- Delivery List Item ---
export interface DeliveryListItem {
  deliveryId: string
  salesOrderId?: string
  deliveryType: string
  deliveryStatus: string
  shipToParty: string
  shipToPartyName?: string
  plannedDeliveryDate?: string
  plannedGiDate?: string
  actualGiDate?: string
  pickingDate?: string
  shippingPoint?: string
  totalQuantity?: number
  deliveredQuantity?: number
}

// --- Pagination ---
export interface Pagination {
  page: number
  pageSize: number
  total: number
  totalPages: number
}

// --- Delivery List Response ---
export interface DeliveryListResponse {
  items: DeliveryListItem[]
  pagination: Pagination
}

// --- Goods Issue ---
export interface GoodsIssue {
  goodsIssueId: string
  deliveryItemId: string
  actualQuantity: number
  postingDate: string
  goodsIssueTime: string
  warehouse?: string
  materialId?: string
  materialName?: string
  batchNo: number
}


// --- Partial Picking / PGI ---
export interface PartialItemQuantity {
  deliveryItemId: string
  quantity: number
}

export interface PartialPickingRequest {
  items: PartialItemQuantity[]
}

export interface PartialPgiRequest {
  items: PartialItemQuantity[]
}

// --- SO Remaining Quantities ---
export interface SoItemRemaining {
  soItemId: string
  materialId: string
  materialName?: string
  orderQuantity: number
  alreadyDelivered: number
  remainingQuantity: number
  salesUnit?: string
}


// --- Pick Record (batch picking) ---
export interface PickRecord {
  pickId: string
  deliveryItemId: string
  batchNo: number
  pickQuantity: number
  storageLocation?: string
  pickDate: string
  pickedBy?: string
  materialId?: string
  materialName?: string
}

export interface PickBatchRequest {
  items: PartialItemQuantity[]
  storageLocation?: string
  pickedBy?: string
}


export interface StorageLocation {
  slocId: string
  slocName: string
  plant: string
  warehouseNo?: string
  storageType?: string
  storageBin?: string
  description?: string
}
// --- Status Labels ---
export interface StatusLabels {
  [key: string]: string
}

// === API Functions ===

const BASE = '/api/v1/logistics'

export function fetchDeliveries(params?: {
  page?: number
  pageSize?: number
  deliveryNo?: string
  salesOrderNo?: string
  customerName?: string
  status?: string
}) {
  return get<DeliveryListResponse>(`${BASE}/deliveries`, { params })
}

export function fetchDeliveryById(id: string) {
  return get<Delivery>(`${BASE}/deliveries/${id}`)
}

export function createDeliveryFromSalesOrder(soId: string) {
  return post<Delivery>(`${BASE}/deliveries/from-so/${soId}`)
}

export function startPicking(deliveryId: string) {
  return post<Delivery>(`${BASE}/deliveries/${deliveryId}/start-picking`)
}

export function confirmPicking(deliveryId: string) {
  return post<Delivery>(`${BASE}/deliveries/${deliveryId}/confirm-picking`)
}

export function shipDelivery(deliveryId: string) {
  return post<Delivery>(`${BASE}/deliveries/${deliveryId}/ship`)
}

export function postGoodsIssue(deliveryId: string, body?: PartialPgiRequest) {
  return post<Delivery>(`${BASE}/deliveries/${deliveryId}/pgi`, body)
}

export function fetchGoodsIssues(deliveryId: string) {
  return get<GoodsIssue[]>(`${BASE}/deliveries/${deliveryId}/goods-issues`)
}

export function pickBatch(deliveryId: string, body: PickBatchRequest) {
  return post<Delivery>(`${BASE}/deliveries/${deliveryId}/pick-batch`, body)
}

export function fetchPickRecords(deliveryId: string) {
  return get<PickRecord[]>(`${BASE}/deliveries/${deliveryId}/pick-records`)
}

export function fetchSoRemaining(soId: string) {
  return get<SoItemRemaining[]>(`${BASE}/sales-orders/${soId}/remaining-quantities`)
}

export function fetchStorageLocations(params?: { keyword?: string; plant?: string }) {
  return get<StorageLocation[]>(`${BASE}/storage-locations`, { params })
}

export function fetchStatusLabels() {
  return get<StatusLabels>(`${BASE}/delivery-status-labels`)
}
